File: Recursion-Backtracking/04-Combination-Sum/combination-sum.cpp
class Solution {
public:

    void findCombinations(int index, int n, int curr_sum, vector<int>& candidates, int target, vector<int>& curr, set<vector<int>>& s){
        if(index == n) return;
        if(curr_sum > target) return;

        if(curr_sum == target){
            s.insert(curr);
            return;
        }

        curr.push_back(candidates[index]);
        findCombinations(index, candidates.size(), curr_sum + candidates[index], candidates, target, curr, s);
        curr.pop_back();
        findCombinations(index + 1, candidates.size(), curr_sum, candidates, target, curr, s);
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        set<vector<int>> s;
        vector<int> curr;
        findCombinations(0, candidates.size(), 0, candidates, target, curr, s);
        vector<vector<int>> result(s.begin(), s.end());
        return result;
    }
};

File: Recursion-Backtracking/04-Combination-Sum/combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
function findCombinations(index, n, curr_sum, target, curr, result, m, candidates){
    if(index === n) return;
    if(curr_sum > target) return;

    if(curr_sum === target){
        const sortedArr = [...curr].sort((a,b) => a - b)
        let key = JSON.stringify(sortedArr);
        if(!m.hasOwnProperty(key)){
            result.push([...curr]);
            m[key] = true;
        }
        return;
    }

    curr.push(candidates[index]);
    findCombinations(index, n, curr_sum + candidates[index], target, curr, result, m, candidates);
    curr.pop();
    findCombinations(index + 1, n, curr_sum, target, curr, result, m, candidates);
}

var combinationSum = function(candidates, target) {
    let m = {};
    let result = [];
    let curr = [];
    findCombinations(0, candidates.length, 0, target, curr, result, m, candidates);
    return result;
};